jax_galsim/moffat.py

```diff
@@ -0,0 +1,476 @@
+        if R < R5hlr:
+            R = R5hlr
+        return jnp.pi / R
+
+    @property
+    def _stepk(self):
+        """The fractional flux out to radius R is (if not truncated)
+        1 - (1+(R/rd)^2)^(1-beta)
+        So solve (1+(R/rd)^2)^(1-beta) = folding_threshold
+        """
+        return jax.lax.select(
+            self.beta <= self._beta_thr, self._stepk_lowbeta, self._stepk_highbeta
+        )
+
+    @property
+    def _has_hard_edges(self):
+        return self.trunc != 0.0
+
+    @property
+    def _max_sb(self):
+        return self._norm
+
+    def _xValue(self, pos):
+        rsq = (pos.x**2 + pos.y**2) * self._inv_r0_sq
+        # trunc if r>maxR with r0 scaled version
+        return jax.lax.select(
+            rsq > self._maxRrD_sq, 0.0, self._norm * jnp.power(1.0 + rsq, -self.beta)
+        )
+
+    def _kValue_untrunc(self, kpos):
+        """Non truncated version of _kValue"""
+        k = jnp.sqrt((kpos.x**2 + kpos.y**2) * self._r0_sq)
+        # return jax.lax.select(
+        #    k == 0,
+        #    self._knorm,
+        #    self._knorm_bis * jnp.power(k, self.beta - 1.0) * _Knu(self.beta - 1, k),
+        # )
+        return jnp.select(
+            [k > 0],
+            [self._knorm_bis * jnp.power(k, self.beta - 1.0) * _Knu(self.beta - 1, k)],
+            default=self._knorm,
+        )
+
+    def _kvalue_trunc(self, kpos):
+        """truncated version of _kValue"""
+        ksq = (kpos.x**2 + kpos.y**2) * self._r0_sq
+        k = jnp.sqrt(ksq)
+        # return jax.lax.select(k > 50.0, 0.0, self._knorm * self._hankel(k))
+        _hankel = lambda x: partial(
+            _xMoffatIntegrant,
+            beta=self.beta,
+            rmax=self._maxRrD,
+            quad=ClenshawCurtisQuad.init(150),
+        )(x)
+        return jnp.select(
+            [k <= 50.0], [self._knorm * self._prefactor * _hankel(k)], default=0.0
+        )
+
+    def _kValue(self, kpos):
+        return jax.lax.cond(
+            self.trunc > 0,
+            lambda x: self._kvalue_trunc(x),
+            lambda x: self._kValue_untrunc(x),
+            kpos,
+        )
+
+    def _drawReal(self, image, jac=None, offset=(0.0, 0.0), flux_scaling=1.0):
+        _jac = jnp.eye(2) if jac is None else jac
+        return draw_by_xValue(self, image, _jac, jnp.asarray(offset), flux_scaling)
+
+    def _drawKImage(self, image, jac=None):
+        _jac = jnp.eye(2) if jac is None else jac
+        return draw_by_kValue(self, image, _jac)
+
+    def withFlux(self, flux):
+        return Moffat(
+            beta=self.beta,
+            scale_radius=self.scale_radius,
+            trunc=self.trunc,
+            flux=flux,
+            gsparams=self.gsparams,
+        )
+
+    # @classmethod
+    # def tree_unflatten(cls, aux_data, children):
+    #    """Recreates an instance of the class from flatten representation"""
+    #    obj = object.__new__(cls)
+    #    return obj
```

Comment:
Did you end up not using this? You might be able to put back some of the checks if you implement the tree flatten and unflatten in a smart way. Let me know if you need help with this @jecampagne

Reply:
but if you got it to work and the checks are not too important we can ignore them for now. Or this is something we can discuss in our next jax-galsim meeting

Reply:
this tree_unflatten commented fucntion is not for the time beeing used. So I delete this part.